glm/detail/func_common_simd.inl

```diff
@@ -225,7 +225,389 @@ namespace detail
+		{
+			vec<4, float, Q> v;
+#if (GLM_ARCH & GLM_ARCH_SSE41_BIT)
+			v.data = _mm_blend_ps(a.data, _mm_setzero_ps(), 8);
+#else
+			__m128i mask = _mm_set_epi32(0, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF);
+			__m128 v0 = _mm_castsi128_ps(_mm_and_si128(_mm_castps_si128(a.data), mask));
+			v.data = v0;
+#endif
+			return v;
+		}
+	};
+
+	// copy vec3 to vec4 and set w to 1
+	template<qualifier Q>
+	struct convert_vec3_to_vec4W1<float, Q, true>
+	{
+		GLM_FUNC_QUALIFIER static vec<4, float, Q> call(vec<3, float, Q> const& a)
+		{
+			vec<4, float, Q> v;
+#if (GLM_ARCH & GLM_ARCH_SSE41_BIT)
+			v.data = _mm_blend_ps(a.data, _mm_set1_ps(1.0f), 8);
+#else
+			__m128 t1 = _mm_shuffle_ps(a.data, a.data, _MM_SHUFFLE(0, 2, 1, 3)); //permute x, w
+			__m128 t2 = _mm_move_ss(t1, _mm_set_ss(1.0f)); // set x to 1.0f
+			v.data = _mm_shuffle_ps(t2, t2, _MM_SHUFFLE(0, 2, 1, 3)); //permute x, w
+#endif
+			return v;
+		}
+	};
+
+	// copy vec3 to vec4 and set w to vec3.z
+	template<qualifier Q>
+	struct convert_vec3_to_vec4WZ<float, Q, true>
+	{
+		GLM_FUNC_QUALIFIER static vec<4, float, Q> call(vec<3, float, Q> const& a)
+		{
+			vec<4, float, Q> v;
+			v.data = _mm_shuffle_ps(a.data, a.data, _MM_SHUFFLE(2, 2, 1, 0));
+			return v;
+		}
+	};
+
+	// copy vec3 to vec4 and set w to 0
+	template<qualifier Q>
+	struct convert_vec3_to_vec4W0<double, Q, true>
+	{
+		GLM_FUNC_QUALIFIER static vec<4, double, Q> call(vec<3, double, Q> const& a)
+		{
+			vec<4, double, Q> v;
+#if (GLM_ARCH & GLM_ARCH_AVX_BIT)
+			v.data = _mm256_blend_pd(a.data, _mm256_setzero_pd(), 8);
+#else
+			v.data.setv(0, a.data.getv(0));
+			glm_dvec2 av2 = a.data.getv(1);
+			av2 = _mm_shuffle_pd(av2, _mm_setzero_pd(), 2);
+			v.data.setv(1, av2);
+#endif
+			return v;
+		}
+	};
+
+	// copy vec3 to vec4 and set w to vec3.z
+	template<qualifier Q>
+	struct convert_vec3_to_vec4WZ<double, Q, true>
+	{
+		GLM_FUNC_QUALIFIER static vec<4, double, Q> call(vec<3, double, Q> const& a)
+		{
+			vec<4, double, Q> v;
+#if (GLM_ARCH & GLM_ARCH_AVX_BIT)
+			v.data = _mm256_permute_pd(a.data, 2);
+#else
+			v.data.setv(0, a.data.getv(0));
+			glm_dvec2 av2 = a.data.getv(1);
+			__m128d t1 = _mm_shuffle_pd(av2, av2, 0);
+			v.data.setv(1, t1);
+#endif
+			return v;
+		}
+	};
+
+	// copy vec3 to vec4 and set w to 1
+	template<qualifier Q>
+	struct convert_vec3_to_vec4W1<double, Q, true>
+	{
+		GLM_FUNC_QUALIFIER static vec<4, double, Q> call(vec<3, double, Q> const& a)
+		{
+			vec<4, double, Q> v;
+#if (GLM_ARCH & GLM_ARCH_AVX_BIT)
+			v.data = _mm256_blend_pd(a.data, _mm256_set1_pd(1.0), 8);
+#else
+			v.data.setv(0, a.data.getv(0));
+			glm_dvec2 av2 = a.data.getv(1);
+			av2 = _mm_shuffle_pd(av2, _mm_set1_pd(1.), 2);
+			v.data.setv(1, av2);
+#endif
+			return v;
+		}
+	};
+
+	template<qualifier Q>
+	struct convert_vec4_to_vec3<float, Q, true> {
+		GLM_FUNC_QUALIFIER static vec<3, float, Q> call(vec<4, float, Q> const& a)
+		{
+			vec<3, float, Q> v;
+			v.data = (glm_f32vec4)a.data;
```

Comment:
On macOS, we are getting an error here:

/Users/runner/work/glm/glm/glm/detail/func_common_simd.inl:[38](https://github.com/g-truc/glm/actions/runs/7397913038/job/20125945709?pr=1212#step:15:39)0:13: error: use of old-style cast [-Werror,-Wold-style-cast]
                        v.data = (glm_f32vec4)a.data;
                                 ^            ~~~~~~
                                 
                                 This also look wrong, if anything if should be `(glm_f32vec3)a.data` ?